mesonbuild/build.py

```diff
@@ -669,6 +669,9 @@ def get_option(self, key: 'OptionKey') -> T.Union[str, int, bool, 'WrapMode']:
         # in the future we might be able to remove the cast here
         return T.cast('T.Union[str, int, bool, WrapMode]', self.options[key].value)
 
+    def get_raw_override(self, key: str) -> T.Optional(str):
```

Comment:
`T.Optional[str]`

Reply:
There's no point in reviewing typing at this point, the methods are probably going to change before this is ready.

Reply:
Yeah, I mostly did a cursory scan to see if there was any design stuff I was concerned about, which there isn't :)

Reply:
There is one design question still unanswered, though. Should options defined in `subproject` and `project` become these additional options? I think they should, because then you can alter them after the fact from the command line without editing source files.

Reply:
Can you give me an example of what you mean?

Reply:
Well suppose you have a subproject that starts with this:

```meson
project('something', default_options: ['cpp_std=c++17'])
```

Should the output of `configure` be something like:

```
<default options here>

Per-project options overrides [or whatever]:

something:cpp_std=c++17
```

And should you be able to unset that so it reverts to the global default with:

```
meson configure -U something:cpp_std
```

Reply:
In my opinion, yes that is what should happen. This implies two different things:

- Any "basic" option should be overriddable
- All such deviations from the default (whether from subproject override, project setting or `meson configure -A`) should be stored in the same way

Note that "override" is a bad term because we already have `override_options` but I did not come up with a better one.

Reply:
On thing that I've always wanted is the ability to override the superproject but not the subproject, so assuming mesa:
```console
meson setup builddir -Ddebug=false -Dmesa:debug=true
```
so that I can easily set a default, but override the main project just like subprojects. Any chance that would work after your changes?

Reply:
That functionality is planned, yes. The actual syntax is not yet final, though. The one I'm currently leaning towards is that this:

```
meson configure -A :cpp_std=c++20
```

would set the option on the top level project. Thus if you want to build your project so that it is in debug mode, but optimisations are enabled on subprojects only you'd do something like:

```
meson setup --buildtype=debug <other args>
meson configure -Doptimization=2
meson configure -A :optimization=0
```

These could be all in one line, I just split it into three for easier readability.